File: files/2026-04-26/CSEC 26042026.R
library(tidyverse)
library(readxl)

path <- "2026-04-26/Challenge 114.xlsx"
input <- read_excel(path, range = "B3:D8")
input2 <- read_excel(path, range = "B10:C12")
test <- read_excel(path, range = "F3:H8")

result <-
  input %>%
  left_join(input2, by = "Customer") %>%
  group_by(Customer) %>%
  arrange(Invoice, .by_group = TRUE) %>%
  mutate(
    cum = cumsum(Amount),
    prev = lag(cum, default = 0),
    Allocated = pmax(0, pmin(Amount, Payment - prev)),
    Remaining = Amount - Allocated
  ) %>%
  ungroup() %>%
  select(Invoice, Allocated, Remaining)

all.equal(result, test)
# [1] TRUE

File: files/2026-04-26/CSEC 26042026.py
import pandas as pd

path = "2026-04-26/Challenge 114.xlsx"
input = pd.read_excel(path, usecols="B:D", nrows=6, skiprows=2)
input2 = pd.read_excel(path, usecols="B:C", nrows=3, skiprows=9)
test = pd.read_excel(path, usecols="F:H", nrows=6, skiprows=2).rename(columns=lambda c: c.replace(".1", ""))

df = (
    input.merge(input2, on="Customer")
    .sort_values(["Customer", "Invoice"])
)
df["cum"] = df.groupby("Customer")["Amount"].cumsum().astype("int64")
df["prev"] = df.groupby("Customer")["cum"].shift(fill_value=0).astype("int64")
df["Allocated"] = (df["Payment"] - df["prev"]).clip(lower=0).clip(upper=df["Amount"]).astype("int64")
df["Remaining"] = (df["Amount"] - df["Allocated"]).astype("int64")
result = df[["Invoice", "Allocated", "Remaining"]]

print(result.equals(test))
# True